frontend/src/metabase/nav/components/SearchBar.unit.spec.tsx

```diff
@@ -0,0 +1,182 @@
+  createMockRecentItem({
+    model_object: createMockModelObject({ name }),
+    model_id: index,
+  }),
+);
+
+const setup = ({
+  initialRoute = "/",
+  searchResultItems = TEST_SEARCH_RESULTS,
+  recentViewsItems = TEST_RECENT_VIEWS_RESULTS,
+} = {}) => {
+  const state = createMockState({
+    settings: createMockSettingsState({
+      "search-typeahead-enabled": true,
+    }),
+  });
+
+  setupSearchEndpoints(searchResultItems);
+  setupRecentViewsEndpoints(recentViewsItems);
+
+  const { history } = renderWithProviders(
+    <Route path="*" component={SearchBar} />,
+    {
+      withRouter: true,
+      initialRoute,
+      storeInitialState: state,
+    },
+  );
+
+  return { history };
+};
+
+const getSearchBar = () => {
+  return screen.getByPlaceholderText("Search…");
+};
+
+describe("SearchBar", () => {
+  describe("typing a search query", () => {
+    it("should change URL when user types a query and hits `Enter`", async () => {
+      const { history } = setup();
+
+      userEvent.type(getSearchBar(), "BC{enter}");
+
+      const location = history?.getCurrentLocation();
+      expect(location?.pathname).toEqual("search");
+      expect(location?.search).toEqual("?q=BC");
+    });
+    it("should render 'No Results Found' when the query has no results", async () => {
+      setup({ searchResultItems: [] });
+      const searchBar = getSearchBar();
+      userEvent.click(searchBar);
+      userEvent.type(searchBar, "XXXXX");
+      await waitFor(() =>
+        expect(screen.queryByTestId("loading-spinner")).not.toBeInTheDocument(),
+      );
+
+      expect(
+        within(screen.getByTestId("search-bar-results-container")).getByText(
+          "Didn't find anything",
+        ),
+      ).toBeInTheDocument();
+    });
+  });
+  describe("focusing on search bar", () => {
+    it("should render `Recent Searches` list when clicking the search bar", async () => {
+      setup();
+      getSearchBar().click();
+      expect(await screen.findByText("Recents ABC")).toBeInTheDocument();
+    });
+
+    it("should render `Nothing here` and a folder icon if there are no recently viewed items", async () => {
+      setup({ recentViewsItems: [] });
+      getSearchBar().click();
+      expect(await screen.findByText("Nothing here")).toBeInTheDocument();
+    });
+  });
+  describe("keyboard navigation", () => {
+    it("should focus on the filter bar when the user tabs from the search bar", () => {
+      setup();
+      getSearchBar().click();
+      userEvent.tab();
+      expect(screen.getByTestId("search-bar-filter-button")).toHaveFocus();
+    });
+    it("should allow navigation through the search results with the keyboard", async () => {
+      setup();
+      getSearchBar().click();
+      userEvent.type(getSearchBar(), "BC");
+
+      await waitFor(() =>
+        expect(screen.queryByTestId("loading-spinner")).not.toBeInTheDocument(),
+      );
+
+      const resultItems = await screen.findAllByTestId("search-result-item");
```

Comment:
After waiting for an element to be removed you should be able to just use `screen.get****` without having to `await` that will make the test more concise. Can you try and see if that works?

Also, if you only renders `test-id="search-result-item"` when there are results from the BE, you should be able to remove `waitForElementToBeRemoved` check since that's redundant.

Reply:
Gotcha, I removed the `waitFor` and kept the `findAllBy` - makes the code look a lot cleaner without the `loading-spinner` logic everywhere

Reply:
Another pattern I saw we do a lot is to `await waitFor(/* wait for loading to disappear */)` in the `setup()` and you use the setup like this `await setup()` then you should be able to use `screen.getBy***` instead of `await screen.findBy***` everywhere.

To be clear, I'm not suggesting you change to this pattern. I just want to say this pattern also works.